server/src/main/java/org/apache/iotdb/db/doublelive/OperationSyncLogService.java

```diff
@@ -0,0 +1,217 @@
+      }
+
+      int nextID;
+      if (firstID == 0) {
+        nextID = logFileIDList.get(logFileIDList.size() - 1) + 1;
+      } else {
+        nextID = logFileIDList.get(firstID - 1) + 1;
+      }
+      logFileID = nextID % maxLogFileNum;
+    } else {
+      logFileID = 0;
+    }
+
+    while (true) {
+      // Check the validity of logFile
+      logWriterLock.lock();
+      try {
+        if (logWriter != null && System.currentTimeMillis() - logFileCreateTime > logFileValidity) {
+          // Submit logFile when it's expired
+          submitLogFile();
+        }
+      } finally {
+        logWriterLock.unlock();
+      }
+
+      try {
+        // Sleep 10s before next check
+        TimeUnit.SECONDS.sleep(10);
+      } catch (InterruptedException e) {
+        LOGGER.error("OperationSyncLogService been interrupted", e);
+      }
+    }
+  }
+
+  private void submitLogFile() {
+    try {
+      logWriter.force();
+    } catch (IOException e) {
+      LOGGER.error("Can't force logWrite", e);
+    }
+    incLogFileSize(logFile.length());
+
+    for (int retry = 0; retry < 5; retry++) {
+      try {
+        logWriter.close();
+      } catch (IOException e) {
+        LOGGER.warn("Can't close OperationSyncLog: {}, retrying...", logFile.getAbsolutePath());
+        try {
+          // Sleep 1s and retry
+          TimeUnit.SECONDS.sleep(1);
+        } catch (InterruptedException ignored) {
+          // Ignore and retry
+        }
+        continue;
+      }
+
+      LOGGER.info("OperationSyncLog: {} is expired and closed", logFile.getAbsolutePath());
+      break;
+    }
+
+    protector.registerLogFile(
+        logFileDir
+            + File.separator
+            + logFileName
+            + (logFileID - 1 + maxLogFileNum) % maxLogFileNum);
+
+    logWriter = null;
+    logFile = null;
+  }
+
+  private void createLogFile() {
+    logFile =
+        SystemFileFactory.INSTANCE.getFile(logFileDir + File.separator + logFileName + logFileID);
+    while (true) {
+      try {
+        if (logFile.createNewFile()) {
+          logFileCreateTime = System.currentTimeMillis();
+          logWriter = new LogWriter(logFile, false);
+          LOGGER.info("Create OperationSyncLog: {}", logFile.getAbsolutePath());
+          break;
+        }
+      } catch (IOException e) {
+        LOGGER.warn("Can't create OperationSyncLog: {}, retrying...", logFile.getAbsolutePath());
+        try {
+          TimeUnit.SECONDS.sleep(1);
+        } catch (InterruptedException ignored) {
+          // Ignore and retry
+        }
+      }
+    }
+    logFileID = (logFileID + 1) % maxLogFileNum;
+  }
+
+  public static synchronized void incLogFileSize(long size) {
+    currentLogFileSize = currentLogFileSize + size;
+  }
+
+  public void acquireLogWriter() {
+    logWriterLock.lock();
+  }
+
+  public void write(ByteBuffer buffer) throws IOException {
+    if (currentLogFileSize < maxLogFileSize) {
```

Comment:
We should write one error log to indicate (currentLogFileSize >= maxLogFileSize)

Reply:
Fixed.